### DISCUSSION TITLE
2FA Login Error on 1.23.1

### DISCUSSION THREAD

**Discussion by emaraszek** [dani-garcia/vaultwarden#2168] [Q&A]
### Subject of the issue

Cannot login as a user with 2FA enabled after updating from 1.23.0 to 1.23.1

### Deployment environment

* vaultwarden version: 2.25.0

* Install method: Docker

* Clients used:  web vault, desktop, iOS

* Reverse proxy and version: Traefik 2.5.5

* MySQL/MariaDB or PostgreSQL version: MariaDB 10.6.5

* Other relevant details:
  * If I use the admin panel to remove 2FA from the account, I am able to login successfully.
  * I created a new container using 1.23.1 with a fresh MariaDB database. I then created a new user and configured 2FA. With this new container and user, I had no issue logging in/entering the TOTP code. 
  * After setting my production container back to 1.23.0, I was able to login again as a user with 2FA enabled.

### Steps to reproduce

1. Update environment from 1.23.0 to 1.23.1 (Docker, MariaDB 10.6.5)
2. Navigate to web vault login page
3. Enter email and password of a user with 2FA enabled (TOTP via authenticator app)
4. Click 'Log In'

### Expected behaviour
User is prompted to enter TOTP code and is logged in successfully after doing so.

### Actual behaviour
Login page displays an error message ("Error adding twofactor_incomplete record") and the container log displays the following:

> "Cannot add or update a child row: a foreign key constraint fails (`bitwarden`.`twofactor_incomplete`, CONSTRAINT `twofactor_incomplete_ibfk_1` FOREIGN KEY (`user_uuid`) REFERENCES `users` (`uuid`))"

### Troubleshooting data

Log file: [_Bitwarden_logs.txt](https://github.com/dani-garcia/vaultwarden/files/7740042/_Bitwarden_logs.txt)

![Screenshot 2021-12-18 030633](https://user-images.githubusercontent.com/7446498/146657247-ca7888d3-bbaa-430c-b767-16a48928ae8a.jpg)

**Comment by ndonkersloot** [comment]
It works for me, i also run vaultwarden in docker.
Are you sure you have the correct time set on the docker host?

You can check this on the admin page in the diagnostics tab. Is "Date & Time (UTC)" marked with "OK"?

**Comment by emaraszek** [comment]
yeah that's marked as "Ok" for me as well
![image](https://user-images.githubusercontent.com/7446498/146850996-1c42cfe6-a9c7-43a2-aa10-792aaff82e94.png)

**Comment by BlackDex** [comment]
I think you need to change your database charset/collation. Please see https://github.com/dani-garcia/vaultwarden/discussions/1492#discussioncomment-544064

**Comment by emaraszek** [comment]
Ok, I just ran through those steps. All tables were converted correctly, except 'invitations'. That produced an error message:

`#1071 - Specified key was too long; max key length is 1000 bytes`

I just recreated the container with 1.23.1 and am still having the same issue when attempting to login. Reverting back to 1.23.0 again for now.

**Comment by jjlin** [comment]
`twofactor_incomplete` is a new table, which you presumably also need to convert.

@BlackDex Maybe you can update your comment, or perhaps just add it to the [wiki](https://github.com/dani-garcia/vaultwarden/wiki/Using-the-MariaDB-%28MySQL%29-Backend).

**Comment by BlackDex** [comment]
Well, those commands are generated actually by the querie above.
So it should only add the tables which are there, it could be that @emaraszek didn't run the `SELECT CONTACT...` query and just copied the example?

But it stated that the key was too long. I think this is because some conversion from what it was, i think latin1 or something, to utf8mb, which is longer indeed.

@emaraszek could you run the following query and show us the output?
```sql
SHOW CREATE TABLE `invitations`; 
```

It should output something like this:
```sql
CREATE TABLE `invitations` (
  `email` varchar(255) NOT NULL,
  PRIMARY KEY (`email`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
```

**Comment by BlackDex** [comment]
btw @jjlin i updated the wiki: https://github.com/dani-garcia/vaultwarden/wiki/Using-the-MariaDB-(MySQL)-Backend#foreign-key-errors-collation-and-charset

**Comment by emaraszek** [comment]
@BlackDex I did run the `SELECT CONTACT...` query as instructed, which did pick up the twofactor_incomplete table. That table appears to have converted correctly:

```
CREATE TABLE `twofactor_incomplete` (
  `user_uuid` char(36) COLLATE utf8mb4_unicode_ci NOT NULL,
  `device_uuid` char(36) COLLATE utf8mb4_unicode_ci NOT NULL,
  `device_name` text COLLATE utf8mb4_unicode_ci NOT NULL,
  `login_time` datetime NOT NULL,
  `ip_address` text COLLATE utf8mb4_unicode_ci NOT NULL,
  PRIMARY KEY (`user_uuid`,`device_uuid`),
  CONSTRAINT `twofactor_incomplete_ibfk_1` FOREIGN KEY (`user_uuid`) REFERENCES `users` (`uuid`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```
However, the invitations table has not been converted and returns the following:
```
CREATE TABLE `invitations` (
  `email` varchar(255) NOT NULL,
  PRIMARY KEY (`email`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1
```
I do have 2 unaccepted invitations pending, so perhaps that could be the cause? However, that seems like a separate issue from the MFA login problem on 1.23.1

**Comment by BlackDex** [comment]
What i find strange is that one table is using innodb and the other myisam. Did you changed database backends or updated the configuration? All should be the same. And they all should be utf8mb.

**Comment by emaraszek** [comment]
@BlackDex Thank you, your last comment highlighted an issue with my environment that I had missed.

I did migrate from SQLite to MariaDB a couple years ago, so these tables must have been using MyISAM since then. I could see that newer tables (twofactor_incomplete, sends, emergency_access) were using InnoDB as expected. I did follow the migration guide in the wiki, so I'm not sure if I missed something or my migrated tables defaulted to MyISAM for some reason.

I did a mass-conversion of all tables to InnoDB using a similar approach to your query:

```
SELECT CONCAT('ALTER TABLE `', TABLE_NAME,'` ENGINE = InnoDB;') AS CharSetConvert
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_SCHEMA="bitwarden"
AND TABLE_TYPE="BASE TABLE";
```
After changing all tables to InnoDB, I was able to set the charset/collation properly using your query. I then recreated the container with 1.23.1 and was able to login successfully. 

Perhaps this should be added to the wiki as well?

**Comment by JeromeGillard** [comment]
Had a similar issue when restoring from backup. The file ownership/permissions were not right.  
Vaultwarden web ui was giving me the error `Error adding twofactor_incomplete record`.
Vaultwarden docker container gave me those (more accurate) log messages:
```
POST /identity/connect/token
Error adding twofactor_incomplete record.
[CAUSE] DatabaseError(
    Unknown,
    "attempt to write a readonly database",
)
```
 A simple `sudo chmod me vaultwarden/data -R` solved my issue.

**Comment by david-nano** [comment]
Same error here. This is the logs from the pod:
```
[2023-10-04 06:36:03.033][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2023-10-04 06:36:12.088][request][INFO] POST /identity/connect/token
[2023-10-04 06:36:12.835][error][ERROR] Error adding twofactor_incomplete record.
[CAUSE] DatabaseError(
    Unknown,
    "disk I/O error",
)
[2023-10-04 06:36:12.835][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2023-10-04 06:36:50.824][panic][ERROR] thread 'tokio-runtime-worker' panicked at 'Error loading twofactor_incomplete: DatabaseError(Unknown, "database disk image is malformed")': src/db/models/two_factor_incomplete.rs:79
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: vaultwarden::db::models::two_factor_incomplete::TwoFactorIncomplete::find_logins_before::{{closure}}::{{closure}}
   8: tokio::runtime::scheduler::multi_thread::worker::block_in_place
   9: vaultwarden::api::core::two_factor::send_incomplete_2fa_notifications::{{closure}}
  10: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  11: tokio::runtime::task::core::Core<T,S>::poll
  12: tokio::runtime::task::harness::Harness<T,S>::poll
  13: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
  14: tokio::runtime::scheduler::multi_thread::worker::Context::run
  15: tokio::macros::scoped_tls::ScopedKey<T>::set
  16: tokio::runtime::scheduler::multi_thread::worker::run
  17: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  18: tokio::runtime::task::core::Core<T,S>::poll
  19: tokio::runtime::task::harness::Harness<T,S>::poll
  20: tokio::runtime::blocking::pool::Inner::run
  21: std::sys_common::backtrace::__rust_begin_short_backtrace
  22: core::ops::function::FnOnce::call_once{{vtable.shim}}
  23: std::sys::unix::thread::Thread::new::thread_start
  24: start_thread
  25: clone
```


Edit:
After removing cookies from site - login is OK.

**Reply by stefan0xC** [reply]
> `DatabaseError(Unknown, "database disk image is malformed")':`
> ` "disk I/O error"`

Sounds like a completely different error. You should probably check if the database is okay. (A corrupted database could also point to a failing hard drive.)
